src/benchmark.cc

```diff
@@ -57,596 +59,277 @@ DEFINE_double(benchmark_min_time, 0.5,
+      // Wait for all threads to enter
+      int phase_number_cp = phase_number_;
+      auto cb = [this, phase_number_cp]() {
+        return this->phase_number_ > phase_number_cp;
+      };
+      phase_condition_.wait(ml.native_handle(), cb);
+      return false;  // I was not the last one
+    } else {
+      // Last thread has reached the barrier
+      phase_number_++;
+      entered_ = 0;
+      return true;
     }
   }
-
-  BENCHMARK_DISALLOW_COPY_AND_ASSIGN(FastClock);
 };
 
-struct State::ThreadStats {
-  int64_t bytes_processed;
-  int64_t items_processed;
-
-  ThreadStats() { Reset(); }
-
-  void Reset() {
-    bytes_processed = 0;
-    items_processed = 0;
-  }
-
-  void Add(const ThreadStats& other) {
-    bytes_processed += other.bytes_processed;
-    items_processed += other.items_processed;
-  }
-};
+// TimerManager for current run.
+static TimerManager* timer_manager = nullptr;
 
 namespace internal {
 
+const int Benchmark::kNumCpuMarker;
+
 // Information kept per benchmark we may want to run
 struct Benchmark::Instance {
-  Instance()
-      : bm(nullptr),
-        threads(1),
-        rangeXset(false),
-        rangeX(kNoRange),
-        rangeYset(false),
-        rangeY(kNoRange) {}
-
-  std::string name;
-  Benchmark* bm;
-  int threads;  // Number of concurrent threads to use
-
-  bool rangeXset;
-  int rangeX;
-  bool rangeYset;
-  int rangeY;
-
-  bool multithreaded() const { return !bm->thread_counts_.empty(); }
+  std::string   name;
+  Function*     function;
+  bool          has_arg1;
+  int           arg1;
+  bool          has_arg2;
+  int           arg2;
+  int           threads;    // Number of concurrent threads to use
+  bool          multithreaded;  // Is benchmark multi-threaded?
 };
 
-}  // end namespace internal
-
-struct State::SharedState {
-  const internal::Benchmark::Instance* instance;
-  std::mutex mu;
-  std::condition_variable cond;
-  int starting;  // Number of threads that have entered STARTING state
-  int stopping;  // Number of threads that have entered STOPPING state
-  int exited;    // Number of threads that have complete exited
-  int threads;   // Number of total threads that are running concurrently
-  ThreadStats stats;
-  std::vector<BenchmarkReporter::Run> runs;  // accumulated runs
-  std::string label;
-
-  explicit SharedState(const internal::Benchmark::Instance* b)
-      : instance(b),
-        starting(0),
-        stopping(0),
-        exited(0),
-        threads(b == nullptr ? 1 : b->threads) { }
-
-  BENCHMARK_DISALLOW_COPY_AND_ASSIGN(SharedState);
-};
-
-namespace internal {
-
-Benchmark::Benchmark(const char* name, BenchmarkFunction f)
-    : name_(name), function_(f) {
-  registration_index_ = BenchmarkFamilies::GetInstance()->AddBenchmark(this);
+Benchmark::Benchmark(const char* name,
+                     Function* f) EXCLUDES(GetBenchmarkLock())
+                    : name_(name), function_(f), arg_count_(-1) {
+  MutexLock l(GetBenchmarkLock());
```

Comment:
Because adding it would have resulted in even more churn. It seems like a good abstraction though. Do you want it in this patch?